### PR TITLE
Build multi-platform Docker images/manifests

### PR DESCRIPTION
**Pull request by symbioquine**
The goal of this change is to make it trivial to run farmOS on the most common ARM architectures without the need for a local `docker build` step.

Unfortunately, the change is fairly involved in order to keep the caching and testing working the same way.

Ideally, we'd be able to do a single `docker buildx build --push --platform COMMA_SEPARATED_PLATFORMS_HERE --tag` command, but that wouldn't work for how we test pull requests and handle creating the various tags conditionally - in addition to making sure our test workflows run against the same images prior to actually pushing them to Docker hub.

Instead, we have to do the more manual strategy of using `docker buildx` to build/cache each platform's image individually, then only at the end do we create the final multi-architecture manifest.

The other thing that makes this change slightly larger is that the existing workflow hard-codes the Docker image name `farmos/farmos` so in order to test that has to be changed in many places. This change has moved that prefix into a variable so it can be changed in a single place to make verifying the changes easier.

References:

* https://www.docker.com/blog/multi-arch-build-and-images-the-simple-way/
* https://github.com/docker/setup-qemu-action
* https://github.com/docker/setup-buildx-action

### CODE REVIEW

**Comment by symbioquine** [comment]
The actions workflow succeeded (with several sqlite retries) for a non-tagged push:

image

I'm now attempting to test whether these changes can successfully build the multi-platform images/manifest for the 3.2.3 release of farmOS.

* https://github.com/symbioquine/farmOS/actions/runs/10228853359
* https://github.com/symbioquine/farmOS/actions/runs/10228853311

image

That is with the following changes pushed to my fork's 3.x branch:

* Change the `DOCKER_IMG_PREFIX` to `symbioquine/farmos`
* Change the check for `github.repository == 'farmOS/farmOS'` to match my fork `symbioquine/farmOS`
* Disable the announcements - since they won't work anyway (I don't have a good way to test these, but I suspect that's an acceptable risk especially since they haven't been changed significantly.)

image

Other than this testing, (assuming it succeeds) the only other things that remain here IMHO are;

* Add a CHANGELOG entry and put the PR id in the commit
* See if it makes sense to merge in @wotnak's caching changes from https://github.com/farmOS/farmOS/pull/764

**Comment by mstenta** [comment]
Great progress @symbioquine!

Bummer that the ARM images take so much longer to build. More than double `amd64`'s build time. :-(

**Comment by symbioquine** [comment]
> Bummer that the ARM images take so much longer to build. More than double `amd64`'s build time. :-(

Yeah, I think we could consider running the ARM builds on ARM runners instead of using QEMU in the future, but I suspect it would be a more involved change.

The other thing is it would be nice if the amd64 tests could happen in parallel with those other platform builds. Again probably a future improvement, but something I'd like to look into. (Not sure if you can do this with separate build matrices...)

**Comment by symbioquine** [comment]
Unfortunately, the 3.2.3 tag build is going to be permanently broken going forward (even if we re-ran the existing workflow it would fail.) This is because we don't lock dependencies and new Composer dependencies have been picked up that break the CodeSniffer rules and the tests.

To get around this, I'm making one more change in my 3.x branch for testing purposes - that is, I'm making it so my tagged release still builds with the 3.x-dev `FARMOS_COMPOSER_VERSION`;

image

By doing that, we can pick up the fixes that have since been made on the main 3.x branch of farmOS.

Here are the resulting actions workflow runs; (still pending)

* https://github.com/symbioquine/farmOS/actions/runs/10229348246
* https://github.com/symbioquine/farmOS/actions/runs/10229348269

**Comment by symbioquine** [comment]
> Yeah, I think we could consider running the ARM builds on ARM runners instead of using QEMU in the future, but I suspect it would be a more involved change.

That might be blocked on https://github.com/github/roadmap/issues/970

**Comment by symbioquine** [comment]
> The other thing is it would be nice if the amd64 tests could happen in parallel with those other platform builds. Again probably a future improvement, but something I'd like to look into. (Not sure if you can do this with separate build matrices...)

Relevant thread for future reference: https://github.com/orgs/community/discussions/42335

**Comment by symbioquine** [comment]
Found something I missed, and force pushed the fix. New test workflow runs;

* https://github.com/symbioquine/farmOS/actions/runs/10229718062
* https://github.com/symbioquine/farmOS/actions/runs/10229718073

**Comment by symbioquine** [comment]
I believe I finally got this working all the way. The latest test workflow runs are;

* https://github.com/symbioquine/farmOS/actions/runs/10250482744
* https://github.com/symbioquine/farmOS/actions/runs/10250482692

image

Note, the failure here is something I messed up in attempting to disable the "announce" steps as part of my [testing commit](https://github.com/symbioquine/farmOS/commit/40575912ac9f7a2dc06a4f60593ef3c55bfff5bf#diff-af086e6040f0bcb375db877f5a696150b5def4c0060b6262bd21918e1592e603L331), but shouldn't be relevant to the part of the workflow I'm trying to test.

Resulting images/manifests pushed:

https://hub.docker.com/repository/docker/symbioquine/farmos/tags

image

I was able to do `docker run --rm -p 80:80 symbioquine/farmos` on a Raspberry Pi 3B+ running 64bit Debian Bookworm (12).

File: .github/workflows/deliver.yml
Comment (symbioquine) [inline .github/workflows/deliver.yml]: Per the Thursday dev call, @mstenta pointed out that these releases should be the same and that we should just do these steps once.

I've confirmed that the resulting releases are identical - with the exception of the local `.git` directory which we're including and which has inconsequential differences (timestamps of checkout events I think):

e.g.

```shell
$ diff -r farmOS-3.2.3-arm32v7 farmOS-3.2.3-arm64v8
Binary files farmOS-3.2.3-arm32v7/.git/index and farmOS-3.2.3-arm64v8/.git/index differ
diff -r farmOS-3.2.3-arm32v7/.git/logs/HEAD farmOS-3.2.3-arm64v8/.git/logs/HEAD
1,3c1,3
< 0000000000000000000000000000000000000000 a7625756734acef0acf968241dc82b7ad0db4d13 root <root@buildkitsandbox.(none)> 1722867565 +0000 clone: from https://github.com/farmOS/composer-project.git
< a7625756734acef0acf968241dc82b7ad0db4d13 a7625756734acef0acf968241dc82b7ad0db4d13 root <root@buildkitsandbox.(none)> 1722867565 +0000 checkout: moving from 3.x to 3.x
< a7625756734acef0acf968241dc82b7ad0db4d13 a7625756734acef0acf968241dc82b7ad0db4d13 root <root@buildkitsandbox.(none)> 1722867565 +0000 reset: moving to HEAD
---
> 0000000000000000000000000000000000000000 a7625756734acef0acf968241dc82b7ad0db4d13 root <root@buildkitsandbox.(none)> 1722867590 +0000 clone: from https://github.com/farmOS/composer-project.git
> a7625756734acef0acf968241dc82b7ad0db4d13 a7625756734acef0acf968241dc82b7ad0db4d13 root <root@buildkitsandbox.(none)> 1722867590 +0000 checkout: moving from 3.x to 3.x
> a7625756734acef0acf968241dc82b7ad0db4d13 a7625756734acef0acf968241dc82b7ad0db4d13 root <root@buildkitsandbox.(none)> 1722867590 +0000 reset: moving to HEAD
diff -r farmOS-3.2.3-arm32v7/.git/logs/refs/heads/3.x farmOS-3.2.3-arm64v8/.git/logs/refs/heads/3.x
1c1
< 0000000000000000000000000000000000000000 a7625756734acef0acf968241dc82b7ad0db4d13 root <root@buildkitsandbox.(none)> 1722867565 +0000 clone: from https://github.com/farmOS/composer-project.git
---
> 0000000000000000000000000000000000000000 a7625756734acef0acf968241dc82b7ad0db4d13 root <root@buildkitsandbox.(none)> 1722867590 +0000 clone: from https://github.com/farmOS/composer-project.git
diff -r farmOS-3.2.3-arm32v7/.git/logs/refs/remotes/origin/HEAD farmOS-3.2.3-arm64v8/.git/logs/refs/remotes/origin/HEAD
1c1
< 0000000000000000000000000000000000000000 a7625756734acef0acf968241dc82b7ad0db4d13 root <root@buildkitsandbox.(none)> 1722867565 +0000 clone: from https://github.com/farmOS/composer-project.git
---
> 0000000000000000000000000000000000000000 a7625756734acef0acf968241dc82b7ad0db4d13 root <root@buildkitsandbox.(none)> 1722867590 +0000 clone: from https://github.com/farmOS/composer-project.git
Binary files farmOS-3.2.3-arm32v7/web/profiles/farm/.git/index and farmOS-3.2.3-arm64v8/web/profiles/farm/.git/index differ
diff -r farmOS-3.2.3-arm32v7/web/profiles/farm/.git/logs/HEAD farmOS-3.2.3-arm64v8/web/profiles/farm/.git/logs/HEAD
1,2c1,2
< 0000000000000000000000000000000000000000 40575912ac9f7a2dc06a4f60593ef3c55bfff5bf root <root@buildkitsandbox.(none)> 1722867648 +0000 clone: from /tmp/tmp.362UYC7utT/cache/vcs/https---github.com-symbioquine-farmOS/
< 40575912ac9f7a2dc06a4f60593ef3c55bfff5bf 40575912ac9f7a2dc06a4f60593ef3c55bfff5bf root <root@buildkitsandbox.(none)> 1722867651 +0000 checkout: moving from 3.x to 40575912ac9f7a2dc06a4f60593ef3c55bfff5bf
---
> 0000000000000000000000000000000000000000 40575912ac9f7a2dc06a4f60593ef3c55bfff5bf root <root@buildkitsandbox.(none)> 1722867686 +0000 clone: from /tmp/tmp.7KFhyI9WcC/cache/vcs/https---github.com-symbioquine-farmOS/
> 40575912ac9f7a2dc06a4f60593ef3c55bfff5bf 40575912ac9f7a2dc06a4f60593ef3c55bfff5bf root <root@buildkitsandbox.(none)> 1722867689 +0000 checkout: moving from 3.x to 40575912ac9f7a2dc06a4f60593ef3c55bfff5bf
diff -r farmOS-3.2.3-arm32v7/web/profiles/farm/.git/logs/refs/heads/3.x farmOS-3.2.3-arm64v8/web/profiles/farm/.git/logs/refs/heads/3.x
1c1
< 0000000000000000000000000000000000000000 40575912ac9f7a2dc06a4f60593ef3c55bfff5bf root <root@buildkitsandbox.(none)> 1722867648 +0000 clone: from /tmp/tmp.362UYC7utT/cache/vcs/https---github.com-symbioquine-farmOS/
---
> 0000000000000000000000000000000000000000 40575912ac9f7a2dc06a4f60593ef3c55bfff5bf root <root@buildkitsandbox.(none)> 1722867686 +0000 clone: from /tmp/tmp.7KFhyI9WcC/cache/vcs/https---github.com-symbioquine-farmOS/
diff -r farmOS-3.2.3-arm32v7/web/profiles/farm/.git/logs/refs/remotes/origin/HEAD farmOS-3.2.3-arm64v8/web/profiles/farm/.git/logs/refs/remotes/origin/HEAD
1c1
< 0000000000000000000000000000000000000000 40575912ac9f7a2dc06a4f60593ef3c55bfff5bf root <root@buildkitsandbox.(none)> 1722867648 +0000 clone: from /tmp/tmp.362UYC7utT/cache/vcs/https---github.com-symbioquine-farmOS/
---
> 0000000000000000000000000000000000000000 40575912ac9f7a2dc06a4f60593ef3c55bfff5bf root <root@buildkitsandbox.(none)> 1722867686 +0000 clone: from /tmp/tmp.7KFhyI9WcC/cache/vcs/https---github.com-symbioquine-farmOS/
diff -r farmOS-3.2.3-arm32v7/web/profiles/farm/.git/objects/info/packs farmOS-3.2.3-arm64v8/web/profiles/farm/.git/objects/info/packs
1c1
< P pack-6947608a5509b77434cf0d3b7acf81f4e9d75744.pack
---
> P pack-1a2be076db4ddbb5c5cccfdb4265a0b86a5bab27.pack
Only in farmOS-3.2.3-arm64v8/web/profiles/farm/.git/objects/pack: pack-1a2be076db4ddbb5c5cccfdb4265a0b86a5bab27.idx
Only in farmOS-3.2.3-arm64v8/web/profiles/farm/.git/objects/pack: pack-1a2be076db4ddbb5c5cccfdb4265a0b86a5bab27.pack
Only in farmOS-3.2.3-arm32v7/web/profiles/farm/.git/objects/pack: pack-6947608a5509b77434cf0d3b7acf81f4e9d75744.idx
Only in farmOS-3.2.3-arm32v7/web/profiles/farm/.git/objects/pack: pack-6947608a5509b77434cf0d3b7acf81f4e9d75744.pack
```

File: .github/workflows/deliver.yml
Comment (symbioquine) [inline .github/workflows/deliver.yml]: Per feedback on Thursday dev call, I'll pull this out into a separate commit.

**Comment by symbioquine** [comment]
Moved this back to draft since I'm still working on a bit of cleanup relative to the rebase against https://github.com/farmOS/farmOS/pull/866

**Comment by symbioquine** [comment]
I've updated this to be based on https://github.com/farmOS/farmOS/pull/866 and to only publish the release tar.gz once - also to add a few more comments and naming consistency fixes.

The new testing workflow runs are;

* https://github.com/symbioquine/farmOS/actions/runs/10352958199
* https://github.com/symbioquine/farmOS/actions/runs/10352958264

Since this is based on another unmerged PR, the commit to review here is just: https://github.com/farmOS/farmOS/pull/864/commits/ba37f1a81cceefe1ebbbc4d13974e212e2af55bc

**Review by mstenta — APPROVED** [review]

**Review by paul121 — APPROVED** [review]

**Comment by mstenta** [comment]
Rebased onto 3.x (after merging #866), and removed the last paragraph from the commit message (since that bit was refactored out to #866). This is ready to merge! I'll wait until we are ready to tag v3.3.0, and we'll include a more detailed announcement in the blog post. This is a great step @symbioquine - thanks again!!

**Pull request merged**
